_posts/2016-05-31-unit-vs-integration-akka-testing.md

```diff
@@ -0,0 +1,143 @@
+  import akka.pattern.ask
+  import scala.concurrent.duration._
+  implicit val timeout = Timeout(5 seconds)
+
+  val actorRef = TestActorRef(new EchoActor)
+
+  val future = actorRef ? "hello world"
+  val Success(result: String) = future.value.get
+  result should be("hello world")
+}
+```
+
+Let's see how we can test this, and other scenarios, with an integration testing style in Akka.
+
+## Akka Integration testing
+
+Akka provides `TestKit` class for integration testing. Let's see one of our tests written using that class:
+
+```scala
+class ItemFSMSpec() extends TestKit(ActorSystem("ItemFSMSpec")) with ImplicitSender
+
+"send complete message to the original sender when one item deleted is received and there are not more pending" in {
+    val worker = TestFSMRef(new ItemFSM(itemReportedProducer, itemDeletedBus))
+    worker ! ItemReported(itemId)
+
+    itemDeletedBus.publish(MsgEnvelope(item.partitionKey, ItemDeleted(item)))
+
+    within(200 millis) {
+      expectMsg(Result(Right()))
+    }
+ }
+```
+
+In this particular test we're interested in the message that the FSM is dispatching to the sender. Let's read carefully this line again:
+
+```scala
+worker ! ItemReported(itemId)
+```
+
+Here we're saying: send a message of `ItemReported` type to the actor assigned in the val `worker`. But who is sending that message? Extending and mixing `TestKit` and `ImplicitSender` creates a `testActor` that will be the messages sender. `TestKit` exposes some methods like `expectMsg` to inspect the mailbox of that `testActor`. `within` acts like [Scalatest `eventually`](http://www.artima.com/docs-scalatest-2.0.M5/org/scalatest/concurrent/Eventually.html) but with more power. For instance, as the documentation says:
+
+> It should be noted that if the last message-receiving assertion of the block is expectNoMsg or receiveWhile, the final check of the within is skipped in order to avoid false positives due to wake-up latencies. This means that while individual contained assertions still use the maximum time bound, the overall block may take arbitrarily longer in this case.
+
+Another interesting class is `TestProbe`. If we have several actors in our integration test and we want to verify different messages, using a single `testActor` might be confusing. Even with a single actor, using a `TestProbe` improves readability in some cases:
+
+```scala
+"flush a FSM when it receives a Failed message" in {
+  val fsmProbe = TestProbe()
+  val actorFactory: (ActorContext, ActorRef) => ActorRef = (context, self) => fsmProbe.ref
+  val coordinator = TestActorRef(ItemReportedCoordinator.props(actorFactory))
+
+  fsmProbe.send(coordinator, Result(Left(Exception("Some exception message"))))
+
+  fsmProbe.expectMsg(FlushItemFSM)
+}
+```
+
+In [the previous post](http://codurance.com/2016/05/16/coordination-in-akka/) we introduce that actor factory in order to create an actor pool. In here the `TestProbe` makes more clear who is sending and expecting the messages.
```

Comment:
we introduce that actor factory in order to create an actor pool ==> we introduced the actor factory in order to create an actor pool